Prevent duplicate symbol for dsa_add_new_assertion_failure

r-barnes opened the pull request:
`dsa_add_new_assertion_failure` is currently causing duplicate definition issues. Possible solutions:
1. Put the device code in a .cu file - requires device linking, which would be very painful to get setup.
2. inline the code - could cause bloat, especially since a function might include many DSAs.
3. Anonymous namespace - balances the above two. Putting the code in a .cu file would ensure that there's a single copy of the function, but it's hard to setup. Inlining the code would cause bloat. An anonymous namespace is easy to setup and produces a single copy of the function per translation unit, which allows the function to be called many times without bloat.

Differential Revision: D42998295

ezyang commented:
The compiler is not obligated to inline at every call site. It can still generate a symbol for it (which will get deduplicated.)

r-barnes commented:
@ezyang - Are you saying that you would prefer the use of `inline` over the anonymous namespace? I appreciate your comment, but it doesn't explicitly suggest an action to move this forward.

ezyang commented:
Yes, use inline instead of anonymous namespace. If you are extremely concerned about bloat, explicitly mark it noinline too https://devblogs.microsoft.com/oldnewthing/20200521-00/?p=103777

ezyang reviewed (approved):
I'll go ahead and preemptively approve this pending comments

malfet reviewed (commented):
> `dsa_add_new_assertion_failure` is currently causing duplicate definition issues. Possible solutions:

Not sure I understand how. As you've said in `1` we don't have device linking right now and `__device__` functions should not add result in any CPU symbols, so this is not a problem

And although anonymous namespace works, why not add `static` decorator to `dsa_add_new_assertion_failure` (or remove one from `dstrcpy` that is now wrapped into anonymous namespace)

r-barnes committed:
Prevent duplicate symbol for dsa_add_new_assertion_failure (#94064)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/94064

Test Plan: Sandcastle

Differential Revision: D42998295

fbshipit-source-id: 40518158ccfed21120ab7f420fc5585bbf6188fb

r-barnes commented:
Switched to static; that's a good suggestion.
